src/env.rs

```diff
@@ -51,11 +52,33 @@ lazy_static! {
             .read_to_string(&mut hosts)
             .expect("Unable to read the file");
         let hosts: Hosts = toml::from_str(&hosts).expect("unable to process the hosts.conf file");
-        let master_address = hosts.master.clone();
-        let master_ip = master_address.split(":").collect::<Vec<_>>()[0];
-        let master_port = master_address.split(":").collect::<Vec<_>>()[1]
+        let master_ip: Ipv4Addr = hosts.master.split(":").collect::<Vec<_>>()[0].parse().unwrap();
+        let master_port = hosts.master.split(":").collect::<Vec<_>>()[1]
             .parse()
             .unwrap();
-        Env::new(master_ip.to_string(), master_port)
+        Env::new(master_ip, master_port)
     };
+    pub static ref local_ip: Ipv4Addr = std::env::var("SPARK_LOCAL_IP").unwrap_or("0.0.0.0".to_string()).parse().unwrap();
```

Comment:
SPARK_LOCAL_IP is indeed the right config property to get the local IP address. But when it is not set, we should not set "0.0.0.0" as the default address. This address will be sent to sent to master as a string for it to fetch data from this worker(This can be improved though, later). We should try to find the local IP address by looping over network interfaces and throw an error when we couldn't, which is what Spark does I believe. For now, we can simply panic here unless you are willing to implement that too.  Remaining all changes look good.

Reply:
Ah, I see. I'll have a quick look in to how Spark handles it, but if it's complex or some kind of retry on failure loop I'll just make it panic as you say. I wouldn't want to guess the interface if there are multiple physical or virtual to choose from.

Reply:
sure

Reply:
Spark uses SPARK_LOCAL_IP if it is set, otherwise it uses the first address (in OS-provided order) which isn't local or loop back.

There are lots of crates for getting local addresses, but they're almost all hacks. I saw a couple of good ones which used FFI, but they were implemented for a single platform, so for now it's better to just require configuration.